openedx/core/djangoapps/schedules/tracking.py

```diff
@@ -0,0 +1,109 @@
+
+
+@attr.s
+class CampaignTrackingInfo(object):
+    """
+    A struct for storing the set of UTM parameters that are recognized by tracking tools when included in URLs.
+    """
+    source = attr.ib(default=DEFAULT_CAMPAIGN_SOURCE)
+    medium = attr.ib(default=DEFAULT_CAMPAIGN_MEDIUM)
+    campaign = attr.ib(default=None)
+    term = attr.ib(default=None)
+    content = attr.ib(default=None)
+
+    def to_query_string(self, existing_query_string=None):
+        """
+        Generate a query string that includes the tracking parameters in addition to any existing parameters.
+
+        Note that any existing UTM parameters will be overridden by the values in this instance of CampaignTrackingInfo.
+
+        Args:
+            existing_query_string (str): An existing query string that needs to be updated to include this tracking
+                information.
+
+        Returns:
+            str: The URL encoded string that should be used as the query string in the URL.
+        """
+        parameters = {}
+        if existing_query_string is not None:
+            parameters = parse_qs(existing_query_string)
+
+        for attribute, value in attr.asdict(self).iteritems():
+            if value is not None:
+                parameters['utm_' + attribute] = [value]
+        return urlencode(parameters, doseq=True)
+
+
+@attr.s
+class GoogleAnalyticsTrackingPixel(object):
+    """
+    Implementation of the Google Analytics measurement protocol for email tracking.
+
+    See this document for more info: https://developers.google.com/analytics/devguides/collection/protocol/v1/email
+    """
+    ANONYMOUS_USER_CLIENT_ID = 555
+
+    site = attr.ib(default=None)
+    course_id = attr.ib(default=None)
+
+    version = attr.ib(default=1, metadata={'param_name': 'v'})
+    hit_type = attr.ib(default='event', metadata={'param_name': 't'})
+
+    campaign_source = attr.ib(default=DEFAULT_CAMPAIGN_SOURCE, metadata={'param_name': 'cs'})
+    campaign_medium = attr.ib(default=DEFAULT_CAMPAIGN_MEDIUM, metadata={'param_name': 'cm'})
+    campaign_name = attr.ib(default=None, metadata={'param_name': 'cn'})
+    campaign_content = attr.ib(default=None, metadata={'param_name': 'cc'})
+
+    event_category = attr.ib(default='email', metadata={'param_name': 'ec'})
+    event_action = attr.ib(default='edx.bi.email.opened', metadata={'param_name': 'ea'})
+    event_label = attr.ib(default=None, metadata={'param_name': 'el'})
+
+    document_path = attr.ib(default=None, metadata={'param_name': 'dp'})
+
+    user_id = attr.ib(default=None, metadata={'param_name': 'uid'})
+    client_id = attr.ib(default=ANONYMOUS_USER_CLIENT_ID, metadata={'param_name': 'cid'})
+
+    @property
+    def image_url(self):
+        """
+        A URL to a clear image that can be embedded in HTML documents to track email open events.
+
+        The query string of this URL is used to capture data about the email and visitor.
+        """
+        parameters = {}
```

Comment:
Are there any (non-defaulting) parameters that should always be there - and we should raise an error if they aren't provided?  For example, `campaign_name`, `campaign_content`, `document_path`?  Add explicit validators for them?

Reply:
I don't think so? All of the required parameters are defaulted.